File: sdk/root_subnet.py
"""
Root Subnet (Subnet 0) Management

This module implements the Root Subnet functionality for ModernTensor,
managing all registered subnets and root validator weight voting.

Synced with Luxtensor Rust backend:
- luxtensor-core/src/subnet.rs
- luxtensor-node/src/root_subnet.rs
"""

import logging
from typing import Dict, List, Optional
from dataclasses import dataclass, field

from .models.subnet import (
    SubnetInfo,
    RootConfig,
    RootValidatorInfo,
    SubnetWeights,
    EmissionShare,
    SubnetRegistrationResult,
)

logger = logging.getLogger(__name__)


@dataclass
class RootSubnet:
    """
    Root Subnet (Subnet 0) state management.

    The Root Subnet manages:
    - All registered subnets (netuid 1, 2, 3, ...)
    - Root validators (top 64 stakers)
    - Weight voting for emission distribution
    - Emission share calculation

    Synced with Rust: luxtensor-node/src/root_subnet.rs::RootSubnet
    """

    # Registered subnets: netuid -> SubnetInfo
    subnets: Dict[int, SubnetInfo] = field(default_factory=dict)

    # Root validators (auto-selected from top stakers)
    root_validators: List[RootValidatorInfo] = field(default_factory=list)

    # Weight votes: validator_address -> SubnetWeights
    weight_matrix: Dict[str, SubnetWeights] = field(default_factory=dict)

    # Computed emission shares: netuid -> share (0.0 - 1.0)
    emission_shares: Dict[int, float] = field(default_factory=dict)

    # Configuration
    config: RootConfig = field(default_factory=RootConfig)

    # Current state
    next_netuid: int = 1  # Next available subnet ID
    last_weight_update_block: int = 0

    def register_subnet(
        self,
        name: str,
        owner: str,
        block_number: int,
        metadata: Optional[Dict] = None
    ) -> SubnetRegistrationResult:
        """
        Register a new subnet.

        Args:
            name: Human-readable subnet name
            owner: Owner address
            block_number: Current block number
            metadata: Optional metadata dict

        Returns:
            SubnetRegistrationResult with success status and assigned netuid
        """
        # Check max subnets
        if len(self.subnets) >= self.config.max_subnets:
            return SubnetRegistrationResult(
                success=False,
                error=f"Maximum subnets ({self.config.max_subnets}) reached"
            )

        # Assign netuid
        netuid = self.next_netuid
        self.next_netuid += 1

        # Create subnet info
        subnet = SubnetInfo(
            id=netuid,
            netuid=netuid,
            subnet_uid=netuid,
            name=name,
            owner=owner,
            created_at=block_number,
            active=True,
        )

        # Register
        self.subnets[netuid] = subnet

        # Initialize emission share to 0
        self.emission_shares[netuid] = 0.0

        logger.info(f"Registered subnet {netuid}: {name} (owner: {owner})")

        return SubnetRegistrationResult(
            success=True,
            netuid=netuid,
            cost_burned=self.config.subnet_registration_cost
        )

    def deregister_subnet(self, netuid: int, caller: str) -> bool:
        """
        Deregister a subnet.

        Args:
            netuid: Subnet ID to deregister
            caller: Address attempting deregistration

        Returns:
            True if successful
        """
        if netuid not in self.subnets:
            logger.warning(f"Subnet {netuid} not found")
            return False

        subnet = self.subnets[netuid]
        if subnet.owner != caller:
            logger.warning(f"Caller {caller} is not owner of subnet {netuid}")
            return False

        # Remove subnet
        del self.subnets[netuid]
        if netuid in self.emission_shares:
            del self.emission_shares[netuid]

        # Remove from weight matrix
        for weights in self.weight_matrix.values():
            if netuid in weights.weights:
                del weights.weights[netuid]

        logger.info(f"Deregistered subnet {netuid}")
        return True

    def update_root_validators(self, stakers: List[tuple]) -> None:
        """
        Update root validators based on top stakers.

        Args:
            stakers: List of (address, stake) tuples sorted by stake DESC
        """
        # Take top N stakers
        top_stakers = stakers[:self.config.max_root_validators]

        # Filter by minimum stake
        self.root_validators = []
        for rank, (address, stake) in enumerate(top_stakers, 1):
            if stake >= self.config.min_stake_for_root:
                self.root_validators.append(RootValidatorInfo(
                    address=address,
                    stake=stake,
                    rank=rank,
                    is_active=True
                ))

        logger.info(f"Updated root validators: {len(self.root_validators)} validators")

    def is_root_validator(self, address: str) -> bool:
        """Check if address is a root validator."""
        return any(v.address.lower() == address.lower() for v in self.root_validators)

    def set_weights(
        self,
        validator: str,
        weights: Dict[int, float],
        block_number: int
    ) -> bool:
        """
        Set subnet weights for a root validator.

        Args:
            validator: Validator address
            weights: Dict of netuid -> weight (must sum to <= 1.0)
            block_number: Current block number

        Returns:
            True if successful
        """
        if not self.is_root_validator(validator):
            logger.warning(f"{validator} is not a root validator")
            return False

        # Validate weights
        if sum(weights.values()) > 1.0 + 1e-6:  # Allow small floating point error
            logger.warning(f"Weights sum to {sum(weights.values())}, must be <= 1.0")
            return False

        # Validate netuids exist
        for netuid in weights.keys():
            if netuid not in self.subnets:
                logger.warning(f"Subnet {netuid} does not exist")
                return False

        # Store weights
        self.weight_matrix[validator] = SubnetWeights(
            validator=validator,
            weights=weights,
            block_updated=block_number
        )

        # Recalculate emission shares
        self._calculate_emission_shares()

        logger.info(f"Updated weights for {validator}: {weights}")
        return True

    def _calculate_emission_shares(self) -> None:
        """
        Calculate emission shares based on stake-weighted votes.

        Formula: share[subnet] = Σ(validator_stake × validator_vote[subnet]) / total_stake
        """
        if not self.root_validators or not self.weight_matrix:
            return

        # Get total stake
        total_stake = sum(v.stake for v in self.root_validators)
        if total_stake == 0:
            return

        # Calculate weighted average
        shares: Dict[int, float] = {}
        for netuid in self.subnets.keys():
            weighted_sum = 0.0
            for validator in self.root_validators:
                if validator.address in self.weight_matrix:
                    vote = self.weight_matrix[validator.address].weights.get(netuid, 0.0)
                    weighted_sum += validator.stake * vote
            shares[netuid] = weighted_sum / total_stake

        # Normalize to sum to 1.0
        total_shares = sum(shares.values())
        if total_shares > 0:
            self.emission_shares = {k: v / total_shares for k, v in shares.items()}
        else:
            self.emission_shares = {k: 0.0 for k in shares.keys()}

    def get_emission_distribution(self, total_emission: int) -> List[EmissionShare]:
        """
        Get emission distribution for all subnets.

        Args:
            total_emission: Total emission amount for this epoch

        Returns:
            List of EmissionShare with amounts
        """
        result = []
        for netuid, share in self.emission_shares.items():
            amount = int(total_emission * share)
            result.append(EmissionShare(
                netuid=netuid,
                share=share,
                amount=amount
            ))
        return result

    def get_subnet(self, netuid: int) -> Optional[SubnetInfo]:
        """Get subnet info by netuid."""
        return self.subnets.get(netuid)

    def get_all_subnets(self) -> List[SubnetInfo]:
        """Get all registered subnets."""
        return list(self.subnets.values())

    def get_root_validator_count(self) -> int:
        """Get number of root validators."""
        return len(self.root_validators)

    def to_dict(self) -> Dict:
        """Serialize to dict."""
        return {
            "subnets": {k: v.model_dump() for k, v in self.subnets.items()},
            "root_validators": [v.model_dump() for v in self.root_validators],
            "weight_matrix": {k: v.model_dump() for k, v in self.weight_matrix.items()},
            "emission_shares": self.emission_shares,
            "config": self.config.model_dump(),
            "next_netuid": self.next_netuid,
        }

File: luxtensor/crates/luxtensor-node/src/root_subnet.rs
//! Root Subnet (Subnet 0) State Management
//!
//! Manages all registered subnets, root validators, and weight voting
//! for emission distribution across the ModernTensor network.
//!
//! Synced with Python SDK: sdk/root_subnet.py

use std::collections::HashMap;
use std::sync::{Arc, RwLock};
use luxtensor_core::{
    SubnetInfo, RootConfig, RootValidatorInfo, SubnetWeights,
    EmissionShare, SubnetRegistrationResult
};
use tracing::{info, warn};

/// Root Subnet state
pub struct RootSubnetState {
    /// Registered subnets: netuid -> SubnetInfo
    pub subnets: HashMap<u16, SubnetInfo>,

    /// Root validators (top stakers)
    pub root_validators: Vec<RootValidatorInfo>,

    /// Weight votes: validator address hex -> SubnetWeights
    pub weight_matrix: HashMap<String, SubnetWeights>,

    /// Computed emission shares: netuid -> share (0-10000 bps)
    pub emission_shares: HashMap<u16, u16>,

    /// Configuration
    pub config: RootConfig,

    /// Next available subnet ID
    next_netuid: u16,

    /// Last weight calculation block
    last_weight_update: u64,
}

impl Default for RootSubnetState {
    fn default() -> Self {
        Self::new()
    }
}

impl RootSubnetState {
    pub fn new() -> Self {
        Self {
            subnets: HashMap::new(),
            root_validators: Vec::new(),
            weight_matrix: HashMap::new(),
            emission_shares: HashMap::new(),
            config: RootConfig::default(),
            next_netuid: 1,
            last_weight_update: 0,
        }
    }

    /// Register a new subnet
    pub fn register_subnet(
        &mut self,
        name: String,
        owner: [u8; 20],
        block_number: u64,
    ) -> SubnetRegistrationResult {
        // Check max subnets
        if self.subnets.len() >= self.config.max_subnets as usize {
            return SubnetRegistrationResult::failure(
                format!("Maximum subnets ({}) reached", self.config.max_subnets)
            );
        }

        // Assign netuid
        let netuid = self.next_netuid;
        self.next_netuid += 1;

        // Create subnet
        let subnet = SubnetInfo::new(netuid, owner, name.clone(), block_number);
        self.subnets.insert(netuid, subnet);

        // Initialize emission share
        self.emission_shares.insert(netuid, 0);

        info!("Registered subnet {}: {} (owner: {:?})", netuid, name, owner);

        SubnetRegistrationResult::success(netuid, self.config.subnet_registration_cost)
    }

    /// Deregister a subnet
    pub fn deregister_subnet(&mut self, netuid: u16, caller: [u8; 20]) -> Result<(), String> {
        let subnet = self.subnets.get(&netuid)
            .ok_or_else(|| format!("Subnet {} not found", netuid))?;

        if subnet.owner != caller {
            return Err("Caller is not subnet owner".to_string());
        }

        self.subnets.remove(&netuid);
        self.emission_shares.remove(&netuid);

        // Remove from weight matrix
        for weights in self.weight_matrix.values_mut() {
            weights.weights.remove(&netuid);
        }

        info!("Deregistered subnet {}", netuid);
        Ok(())
    }

    /// Update root validators from top stakers
    pub fn update_root_validators(&mut self, stakers: Vec<([u8; 20], u128)>) {
        self.root_validators.clear();

        for (rank, (address, stake)) in stakers.iter().enumerate() {
            if rank >= self.config.max_root_validators as usize {
                break;
            }
            if *stake >= self.config.min_stake_for_root {
                self.root_validators.push(RootValidatorInfo::new(
                    *address,
                    *stake,
                    (rank + 1) as u16
                ));
            }
        }

        info!("Updated root validators: {} validators", self.root_validators.len());
    }

    /// Check if address is a root validator
    pub fn is_root_validator(&self, address: &[u8; 20]) -> bool {
        self.root_validators.iter().any(|v| &v.address == address)
    }

    /// Set subnet weights for a validator
    pub fn set_weights(
        &mut self,
        validator: [u8; 20],
        weights: HashMap<u16, u16>,
        block_number: u64,
    ) -> Result<(), String> {
        if !self.is_root_validator(&validator) {
            return Err("Not a root validator".to_string());
        }

        // Validate total <= 10000 bps
        let total: u32 = weights.values().map(|&v| v as u32).sum();
        if total > 10000 {
            return Err(format!("Weights sum {} exceeds 10000", total));
        }

        // Validate netuids exist
        for netuid in weights.keys() {
            if !self.subnets.contains_key(netuid) {
                return Err(format!("Subnet {} does not exist", netuid));
            }
        }

        // Store weights
        let addr_hex = hex::encode(validator);
        let mut subnet_weights = SubnetWeights::new(validator);
        subnet_weights.weights = weights;
        subnet_weights.block_updated = block_number;
        self.weight_matrix.insert(addr_hex.clone(), subnet_weights);

        // Recalculate emissions
        self.calculate_emission_shares();

        info!("Updated weights for validator {}", addr_hex);
        Ok(())
    }

    /// Calculate emission shares based on stake-weighted votes
    fn calculate_emission_shares(&mut self) {
        if self.root_validators.is_empty() || self.weight_matrix.is_empty() {
            return;
        }

        let total_stake: u128 = self.root_validators.iter().map(|v| v.stake).sum();
        if total_stake == 0 {
            return;
        }

        // Calculate weighted sum for each subnet
        let mut shares: HashMap<u16, u128> = HashMap::new();

        for validator in &self.root_validators {
            let addr_hex = hex::encode(validator.address);
            if let Some(weights) = self.weight_matrix.get(&addr_hex) {
                for (netuid, weight_bps) in &weights.weights {
                    let weighted_vote = validator.stake * (*weight_bps as u128);
                    *shares.entry(*netuid).or_insert(0) += weighted_vote;
                }
            }
        }

        // Normalize to basis points
        let total_weighted: u128 = shares.values().sum();
        if total_weighted > 0 {
            for netuid in self.subnets.keys() {
                let weighted = shares.get(netuid).copied().unwrap_or(0);
                let share_bps = ((weighted * 10000) / total_weighted) as u16;
                self.emission_shares.insert(*netuid, share_bps);
            }
        }
    }

    /// Get emission distribution for all subnets
    pub fn get_emission_distribution(&self, total_emission: u128) -> Vec<EmissionShare> {
        self.emission_shares
            .iter()
            .map(|(netuid, share_bps)| {
                let share = *share_bps as f64 / 10000.0;
                let amount = (total_emission as f64 * share) as u128;
                EmissionShare::new(*netuid, share, amount)
            })
            .collect()
    }

    /// Get all subnets
    pub fn get_all_subnets(&self) -> Vec<&SubnetInfo> {
        self.subnets.values().collect()
    }

    /// Get subnet by netuid
    pub fn get_subnet(&self, netuid: u16) -> Option<&SubnetInfo> {
        self.subnets.get(&netuid)
    }

    /// Get root validator count
    pub fn get_root_validator_count(&self) -> usize {
        self.root_validators.len()
    }
}

/// Thread-safe Root Subnet manager
pub type RootSubnet = Arc<RwLock<RootSubnetState>>;

/// Create new Root Subnet instance
pub fn new_root_subnet() -> RootSubnet {
    Arc::new(RwLock::new(RootSubnetState::new()))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_register_subnet() {
        let mut state = RootSubnetState::new();
        let owner = [1u8; 20];

        let result = state.register_subnet("Test Subnet".to_string(), owner, 100);
        assert!(result.success);
        assert_eq!(result.netuid, Some(1));

        let subnet = state.get_subnet(1).unwrap();
        assert_eq!(subnet.name, "Test Subnet");
    }

    #[test]
    fn test_weight_voting() {
        let mut state = RootSubnetState::new();
        let owner = [1u8; 20];
        let validator = [2u8; 20];

        // Register subnet
        state.register_subnet("Test".to_string(), owner, 100);

        // Add validator
        state.root_validators.push(RootValidatorInfo::new(validator, 1000, 1));

        // Set weights
        let mut weights = HashMap::new();
        weights.insert(1u16, 5000u16); // 50%

        let result = state.set_weights(validator, weights, 200);
        assert!(result.is_ok());
    }
}
